File: app/src/main/java/cz/johnyapps/cheers/adapters/recycleradapters/FilterableAdapter.java
package cz.johnyapps.cheers.adapters.recycleradapters;

import android.content.Context;
import android.widget.Filter;
import android.widget.Filterable;

import androidx.annotation.NonNull;
import androidx.annotation.Nullable;

import java.util.ArrayList;
import java.util.List;

import cz.johnyapps.cheers.tools.TextUtils;

public abstract class FilterableAdapter<VIEW_HOLDER extends SelectableAdapter<VIEW_HOLDER, ITEM, List<ITEM>>.SelectableViewHolder, ITEM> extends SelectableAdapter<VIEW_HOLDER, ITEM, List<ITEM>>  implements Filterable {
    @NonNull
    private List<ITEM> items;
    @Nullable
    private List<ITEM> filteredItems = null;

    public FilterableAdapter(@NonNull Context context,
                             @Nullable List<ITEM> items) {
        super(context);
        this.items = items == null ? new ArrayList<>() : items;
    }

    @Override
    protected void onUpdate(@Nullable List<ITEM> items) {
        this.items = items == null ? new ArrayList<>() : items;
        this.filteredItems = null;
    }

    @NonNull
    @Override
    public ITEM getItem(int position) {
        return getShownList().get(position);
    }

    @Override
    public int getItemCount() {
        return getShownList().size();
    }

    @NonNull
    public List<ITEM> getShownList() {
        if (filteredItems == null) {
            return items;
        }

        return filteredItems;
    }

    @Override
    public Filter getFilter() {
        return new Filter() {
            @Override
            protected FilterResults performFiltering(CharSequence constraint) {
                constraint = TextUtils.removeDiacritics(constraint);
                List<ITEM> filtered = new ArrayList<>();

                for (ITEM item : items) {
                    if (filterItem(constraint, item)) {
                        filtered.add(item);
                    }
                }

                FilterResults filterResults = new FilterResults();
                filterResults.values = filtered;
                filterResults.count = filtered.size();
                return filterResults;
            }

            @Override
            protected void publishResults(CharSequence constraint, FilterResults results) {
                filteredItems = (List<ITEM>) results.values;
                notifyDataSetChanged();
            }
        };
    }

    public boolean isFiltered() {
        return filteredItems != null && filteredItems.size() != items.size();
    }

    public void clearFilter() {
        filteredItems = null;
        notifyDataSetChanged();
    }

    protected abstract boolean filterItem(@NonNull CharSequence constraint, @NonNull ITEM item);

    public void filter(@NonNull String query) {
        getFilter().filter(query);
    }
}

File: app/src/main/java/cz/johnyapps/cheers/adapters/FilterableAdapter.kt
package cz.johnyapps.cheers.adapters

import android.widget.Filter
import android.widget.Filterable
import androidx.recyclerview.widget.DiffUtil
import cz.johnyapps.cheers.tools.TextUtils

abstract class FilterableAdapter<T, VH: SelectableAdapter<T, VH>.ViewHolder>(itemCallback: DiffUtil.ItemCallback<T>): ExpandableListAdapter<T, VH>(itemCallback), Filterable {
    abstract fun getItemTextForFilter(item: T): String
    private var ignoreNextSubmit = false
    private var items: List<T> = ArrayList()
    private var filteredItems: List<T>? = null

    fun filterItem(item: T, constraint: CharSequence): Boolean {
        return TextUtils.removeDiacritics(getItemTextForFilter(item).lowercase())
            .toString()
            .contains(constraint)
    }

    fun filter(query: CharSequence) {
        filter.filter(query)
    }

    fun clearFilter() {
        filteredItems = null
        submitList(items)
    }

    fun isFiltered(): Boolean {
        return filteredItems != null
    }

    override fun submitList(list: List<T>?) {
        if (!ignoreNextSubmit) {
            items = list ?: ArrayList()
        }

        super.submitList(list)
    }

    override fun submitList(list: List<T>?, commitCallback: Runnable?) {
        if (!ignoreNextSubmit) {
            items = list ?: ArrayList()
        }

        super.submitList(list, commitCallback)
    }

    override fun getFilter(): Filter {
        return object: Filter() {
            override fun performFiltering(constraint: CharSequence): FilterResults {
                val filterBy = TextUtils.removeDiacritics(constraint.toString().lowercase())
                val filtered: MutableList<T> = ArrayList()

                for (item in items) {
                    if (filterItem(item, filterBy)) {
                        filtered.add(item)
                    }
                }

                val filterResults = FilterResults()
                filterResults.values = filtered
                filterResults.count = filtered.size
                return filterResults
            }

            override fun publishResults(constraint: CharSequence, results: FilterResults) {
                ignoreNextSubmit = true
                filteredItems = results.values as MutableList<T>
                submitList(filteredItems)
            }
        }
    }
}